pkg/relay/ethereum/chain.go

```diff
@@ -300,14 +284,40 @@ func (c *Chain) QueryUnrecievedAcknowledgements(ctx core.QueryContext, seqs []ui
 	return ret, nil
 }
 
-// QueryPacket returns the packet corresponding to a sequence
-func (c *Chain) QueryPacket(ctx core.QueryContext, sequence uint64) (*chantypes.Packet, error) {
-	return c.findPacket(ctx, c.pathEnd.PortID, c.pathEnd.ChannelID, sequence)
-}
+// QueryUnfinalizedRelayedAcknowledgements returns acks and heights that are sent but not received at the latest finalized block on the counterpartychain
+func (c *Chain) QueryUnfinalizedRelayAcknowledgements(ctx core.QueryContext, counterparty core.LightClientICS04Querier) (core.PacketInfoList, error) {
+	checkpoint, err := c.loadCheckpoint(recvCheckpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	packets, err := c.findReceivedPackets(ctx, checkpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	counterpartyHeader, err := counterparty.GetLatestFinalizedHeader()
+	if err != nil {
+		return nil, err
+	}
+
+	counterpartyCtx := core.NewQueryContext(context.TODO(), counterpartyHeader.GetHeight())
+	seqs, err := counterparty.QueryUnreceivedAcknowledgements(counterpartyCtx, packets.ExtractSequenceList())
+	if err != nil {
+		return nil, err
+	}
+
+	packets = packets.Filter(seqs)
+	if len(packets) == 0 {
+		checkpoint = ctx.Height().GetRevisionHeight()
```

Comment:
Since `ctx.Height().GetRevisionHeight()`  is specified as `ToBlock`(it means "end of range"), can we advance one checkpoint further?
```
checkpoint = ctx.Height().GetRevisionHeight() + 1
```

Reply:
Yes. I already knew this but I forget why I didn't do so. I will fix this. Thank you.